deep_qa/training/train_utils.py

```diff
@@ -0,0 +1,113 @@
+        if node_def.op in ['Variable', 'VariableV2']:
+            return variable_device
+        else:
+            return device
+    return _assign
+
+
+def average_gradients(tower_gradients: List[List[Tuple]]):
+    """
+    Given a list of (gradient, variable) pairs from the result of
+    a gradient calculation from multiple GPUs, calculate their
+    average.
+    """
+    # Make a map from variables -> [gradients that are not none].
+    gradient_map = {}
+    for tower in tower_gradients:
+        for grad, variable in tower:
+            if grad is not None:
+                if variable not in gradient_map:
+                    gradient_map[variable] = []
+                gradient_map[variable].append(grad)
+    average_gradient_list = []
+
+    for variable, gradients in gradient_map.items():
+        # variable is a tensor.
+        # gradients is a list of gradients for this tensor to average.
+
+        # Pick any one of the gradients to see if it is an IndexedSlice.
+
+        first_actual_grad = gradients[0]
+        if isinstance(first_actual_grad, tensorflow.IndexedSlices):
+            # If the gradient is type IndexedSlices then this is a sparse
+            # gradient with attributes indices and values. To average, we
+            # need to concat them individually and then create a new
+            # IndexedSlices object.
+            indices = []
+            values = []
+            for grad in gradients:
+                indices.append(grad.indices)
+                values.append(grad.values)
+            all_indices = tensorflow.concat(indices, 0)
+            avg_values = tensorflow.concat(values, 0) / len(gradients)
+            # Deduplicate across indices.
+            values, unique_indicies = _deduplicate_indexed_slices(avg_values, all_indices)
+            sparse_grad = tensorflow.IndexedSlices(values,
+                                                   unique_indicies,
+                                                   dense_shape=first_actual_grad.dense_shape)
+            average_gradient_list.append((sparse_grad, variable))
+        else:
+            # A normal tensor can just do a simple average.
+            grads_expanded = []
+            for grad in gradients:
+                # Add a 0 dimension to the gradients to represent the tower and
+                # append on a 'tower' dimension which we will average over.
+                grads_expanded.append(tensorflow.expand_dims(grad, 0))
+
+            # Average over the 'tower' dimension.
+            grad = tensorflow.concat(grads_expanded, 0)
+            mean_grad = tensorflow.reduce_mean(grad, 0)
+            average_gradient_list.append((mean_grad, variable))
+
+    assert len(average_gradient_list) == len(gradient_map)
+    return average_gradient_list
+
+
+def _deduplicate_indexed_slices(values, indices):
+    """
+    Sums `values` associated with any non-unique `indices`.
```

Comment:
You need double backticks to get monospace font in RST.  Single backticks is italics.